ember-core/src/main/scala/org/http4s/ember/core/Parser.scala

```diff
@@ -141,36 +140,38 @@ private[ember] object Parser {
               state = 1 // set state to check for header value
               name = new String(bv, start, idx - start) // extract name string
               start = idx + 1 // advance past colon for next start
-              if ((bv.size >= idx + 1) && (bv(idx + 1) == space)) {
+
+              // TODO: This if clause may not be necessary since the header value parser trims
+              if ((bv.size > idx + 1 && bv(idx + 1) == space)) {
                 start += 1 // if colon is followed by space advance again
                 idx += 1 // double advance index here to skip the space
               }
               // double CRLF condition - Termination of headers
-            } else if (current == cr && (idx + 1 < bv.size) && (bv(idx + 1) == lf)) {
-              idx += 1 // double advance to drop cr AND lf
+            } else if (current == lf && (idx > 0 && bv(idx - 1) == cr)) {
               complete = true // completed terminate loop
             }
           case 1 => // HeaderValue
             val current = bv(idx)
             // If crlf is next we have completed the header value
-            if (current == cr && ((bv.size >= idx + 1) && bv(idx + 1) == lf)) {
-              val hValue = new String(bv, start, idx - start) // extract header value
+            if (current == lf && (idx > 0 && bv(idx - 1) == cr)) {
```

Comment:
Instead of checking whether the current index is a CR and the next index is a LF, do the opposite; we check if the current index is a LF and the previous index is a CR. This avoids an edge case where a CR is read at the end of a check so an LF isn't available yet, but the index is advanced nonetheless. This means we might do a bit more work when we see a newline, which may be more common in entities